File: functional/service/service.go
package service

import (
	"github.com/yoshiken0927/paradigms/functional/model"
	"github.com/yoshiken0927/paradigms/functional/repository"
)

func PutEmployee(name string) func(int) model.Employee {
	return func(age int) model.Employee {
		return repository.UpdateEmployee(model.SetAge(repository.FindEmployee(name))(age))
	}
}

// func Callback(e model.Employee) func(func(model.Employee)) {
// 	return func(f func(model.Employee)) {
// 		f(e)
// 	}
// }

File: functional/service/Service.hs
module Service
    ( putEmployee,
    ) where

import Model
import Repository

-- TODO バリデーションとエラーハンドリング
putEmployee :: String -> Int -> Employee
putEmployee name age = do
  updateEmployee(setAge (findEmployee name) age)

-- infixl 1 |>
-- (|>) x f = f x
-- putEmployee name age = do
--   (findEmployee name |> setAge $ age) |> updateEmployee
